### PR TITLE
Asynchronous Producer

### PR DESCRIPTION
**Pull request by emmett9001**
This pull request updates the `Producer` to use a thread per broker instead of the main thread for all brokers.

[This code](https://gist.github.com/emmett9001/ef40bac8ce9a160377af) can be used to test this producer's throughput. Here are some [throughput test results with this script](https://gist.github.com/emmett9001/a3a7f0e9f52ac8471912) (how many times per second each producer can return from `produce()`). The old synchronous producer was doing about 4500 messages per second.

Connects to #124.

Todo:
*  [x] Unit test exception raised on worker thread
*  [x] Monitor thread death
*  [x] Abstract thread management logic into the `QueueInfo` class
*  [x] Make `Producer` as small as possible by having it be a special case of `AsyncProducer`
*  [x] Tests

### CODE REVIEW

File: pykafka/producer.py [hunk 1]
```diff
@@ -232,13 +218,27 @@ def _produce(self, message_partition_tups, attempt):
...
+                self._request_built_callback(leader,
+                                             requests.pop(leader.id),
+                                             attempt)
 
         # Send any still not sent
         for leader, req in requests.iteritems():
-            self._send_request(leader, req, attempt)
+            self._request_built_callback(leader, req, attempt)
+
+    def _request_built_callback(self, leader, request, attempt):
```
Comment (emmett9001) [inline pykafka/producer.py]: Bad name, sounds like `request` is a verb.

File: pykafka/producer.py [hunk 2]
```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
...
+        Enqueues the request in the queue associated with `leader`
+
+        :param leader: The broker to which to send the request
+        :type leader: :class:`pykafka.broker.Broker`
+        :param request: The request to send
+        :type request: :class:`pykafka.protocol.ProduceRequest`
+        :param attempt: The attempt number for this send attempt
+        :type attempt: int
+        """
+        self._queues_by_leader[leader.id].put((request, attempt))
```
Comment (emmett9001) [inline pykafka/producer.py]: Should we block here if the queue is too full? That would fix the misleading throughput numbers issue.

Comment (yungchin) [inline pykafka/producer.py]: That sounds useful. I think the java producer has a `queue.buffering.max.messages` config parameter for that - librdkafka has the same.

File: pykafka/producer.py [hunk 3]
```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
...
+        :type broker: :class:`pykafka.broker.Broker`
+        """
+        def worker():
+            while True:
+                time.sleep(.0001)
+                try:
+                    request, attempt = request_queue.get_nowait()
+                except Empty:
+                    continue
+                self._send_request(broker, request, attempt)
```
Comment (yungchin) [inline pykafka/producer.py]: At some value of `attempt` this might raise a `ProduceFailureError` - would that kill the worker?

**Comment by yungchin** [comment]
Hey, cool stuff!

About the error handling in the worker: I guess I've sort of half a feature request in mind while reading this, so there's a hidden agenda :)

The JVM async producer apparently has this option now to confirm delivery of messages, [including their offset in the partition][0], it would be really nice to have something similar. So when catching the `ProduceFailureError` we'd optionally call such an onCompletion callback with the exception, something like that?

The librdkafka producer also has such a delivery callback interface, so there would be no problem on that end.

[0]: http://kafka.apache.org/082/javadoc/org/apache/kafka/clients/producer/Callback.html

**Comment by yungchin** [comment]
Another question: I noticed the ThreadingHandler runs threads in daemon mode - does that mean that if you async-produce some messages, then exit the main thread, there's a chance the worker dies before the messages are sent? (I'm now thinking of something like using `atexit` to check if the queues are empty, but that sounds a bit garish maybe.)

**Comment by yungchin** [comment]
Ok, really cool. I would have thought it would take much more separate code to make an async producer!

So seeing how clean this all is, I just had a thought... I was following the code around where, for `attempt > 0` this is now calling `Producer._produce` from the worker thread (and I was suspicious if that was safe, but it looks like it is all great). So given that that works, actually would it be doable to always have `Producer._produce` run from the worker thread, ie also on the first attempt?

The benefit would be, that then you could do message batching (ie if I'm a user who calls `produce()` with one message at a time, you could still pack a whole bunch of them into one broker request).

Basically, the changes would be something like

 * AsyncProducer would override `_produce()` instead of  `_request_built_callback()`
 * this `_produce()` would stick its `messages` on partition Queues, without building requests for them
 * you'd have the worker threads getting multiple sets of messages off the queue, splicing them together, and calling `Producer._produce` on them for all the request building and sending stuff

Not sure if that makes sense - I need some sleep ;)

File: pykafka/producer.py [hunk 4]
```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
...
         # a random partition distribution changing that on the retry.
         self._produce(self._partition_messages(messages), 0)
+
+
+class AsyncProducer(Producer):
+    """
+    This class implements the asynchronous producer logic found in the
+    JVM driver. In inherits from the synchronous implementation.
+    """
+    def __init__(self, *args, **kwargs):
```
Comment (kbourgoin) [inline pykafka/producer.py]: -1 to not actually having args spelled out. makes it annoying to use if you've got an editor which is doing auto-complete like vim-jedi

File: pykafka/producer.py [hunk 5]
```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
...
+        given broker.
+
+        :param broker: The broker for which to instantiate this queue and worker
+        :type broker: :class:`pykafka.broker.Broker`
+        """
+        def worker():
+            while True:
+                time.sleep(.0001)
+                try:
+                    request, attempt = request_queue.get_nowait()
```
Comment (kbourgoin) [inline pykafka/producer.py]: I don't get why the `nowait` followed by sleep + continue. It's just going to create more busy-waiting like we had with the consumer. The queue is designed for long polling that won't chew through CPU just waiting around.

File: pykafka/producer.py [hunk 6]
```diff
@@ -250,3 +250,68 @@ def produce(self, messages):
...
+                except Empty:
+                    continue
+                self._send_request(broker, request, attempt)
+        request_queue = Queue()
```
Comment (kbourgoin) [inline pykafka/producer.py]: This should use the Queue in `self._cluster.handler` so that we can later have it switch between the `threading` and `gevent` implementations.

File: pykafka/producer.py (hunk 5 above)
Comment (emmett9001) [inline pykafka/producer.py]: So the dequeue operation here should use a timeout? That makes sense.

**Comment by kbourgoin** [comment]
Took a quick look at the implementation and a few things come to mind:

1. You need to monitor thread death. If we're using `threading` (since that's the only handler supported right now) the thread will die silently and you'll just stop producing messages. This is usually solved by creating a registry of the threads running and having a monitor thread which will check their state every so often. Then it can either restart the dead thread or raise the error to the main thread (a tricky thing in its own right).

2. The subclass is very heavily dependent on the implementation of the main class. This isn't technically wrong, but it makes following the implementation of the async producer that much harder.  I'll take another look to puzzle it out, but it feels like a case where trying to avoid duplicated/similar code has made something much more confusing than it needs to be.

3. How does this respond to leaders changing?  If you've got a queue per broker and the broker is no longer correct, will it be shuffled to the right now?  This could be related to point 2, where I don't think I fully understand everything that's going on.

File: pykafka/producer.py (hunk 5 above)
Comment (kbourgoin) [inline pykafka/producer.py]: No, it should block forever and the thread should stay asleep as long as there is no data coming in.  Since it's on a separate thread, is it a problem for it to sit there?

Comment (emmett9001) [inline pykafka/producer.py]: Ah I see - that's not a problem.

File: pykafka/producer.py [hunk 7]
```diff
@@ -241,3 +242,122 @@ def produce(self, messages):
...
+
+        Creates a :class:`queue.Queue` instance and a worker thread for the
+        given broker.
+
+        :param broker: The broker for which to instantiate this queue and worker
+        :type broker: :class:`pykafka.broker.Broker`
+        """
+        def worker():
+            while True:
+                time.sleep(.0001)
```
Comment (kbourgoin) [inline pykafka/producer.py]: Why is there a sleep here? Needs a comment to explain why it's necessary, IMO.

File: pykafka/producer.py [hunk 8]
```diff
@@ -241,3 +242,122 @@ def produce(self, messages):
...
+                try:
+                    request, attempt = request_queue.get_nowait()
+                    self._send_request(broker, request, attempt)
+                except Empty:
+                    continue
+                except ProduceFailureError as e:
+                    log.error("Producer error: %s", e)
```
Comment (kbourgoin) [inline pykafka/producer.py]: This doesn't seem right to me. Why would we log the error and then just continue on? Signaling failure from the thread is annoying, but I think this needs to be bubbled up to the user.

**Comment by kbourgoin** [comment]
Connects to #124.

**Comment by emmett9001** [comment]
Another note for the next time this gets worked on (hopefully tomorrow): @kbourgoin's main complaint about the design of this class so far is that batching across calls to `produce` is not supported. We need to be able to specify a batch size - the size of each message batch sent per request to kafka. This batch will build up in memory across multiple calls to `produce` and will send either when the batch size has been reached or a message in the batch has been waiting for a long enough time.
The current implementation does support batching, but only when all of the messages in the batch were supplied in the same `produce` call. This will be my primary focus when I work on this next, which I hope happens all day tomorrow.

**Comment by emmett9001** [comment]
My latest changes here implement a non-busy waiting scheme similar to the one from @kbourgoin's deque example that uses condition variables to perform signaling between the queue reader thread and the writer thread, which in this case happens to be the main thread.
I'm currently trying to figure out how to handle the leader changing for a partition. Since the queue for that broker will probably be full of messages when this happens, we'll need to find all of the messages in that queue that are for that partition and move them to the queue for that partition's leader.

File: pykafka/producer.py [hunk 9]
```diff
@@ -18,16 +18,17 @@
 limitations under the License.
 """
 __all__ = ["Producer", "AsyncProducer"]
+from collections import defaultdict, deque, namedtuple
 import itertools
 import logging
+import threading
```
Comment (kbourgoin) [inline pykafka/producer.py]: Don't use `threading`. The handler exists to satisfy all of these objects while allowing us to swap out the implementation when we work on gevent compatibility. (e.g. `handler.Lock`)

File: pykafka/producer.py [hunk 10]
```diff
@@ -120,6 +105,11 @@ def _send_request(self, broker, req, attempt):
         :type req: :class:`pykafka.protocol.ProduceRequest`
         :param attempt: The current attempt count. Used for retry logic
         :type attempt: int
+        :param q_info: A QueueInfo instance containing information
+            about the queue from which the messages in `req` were removed.
+            Only used when called from the
+            :class:`pykafka.producer.AsyncProducer`
```
Comment (kbourgoin) [inline pykafka/producer.py]: If this is only relevant when used from `AsyncProducer` why is it in `Producer`?  This smells like something where Async should just have its own implementation instead of trying super hard to share code.

File: pykafka/producer.py [hunk 11]
```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
...
+                lock,
+                flush_ready,
+                slot_available,
+                queue,
+                producer,
+                messages_inflight):
+        ins = super(QueueInfo, cls).__new__(
+            cls, lock, flush_ready, slot_available, queue, producer)
+        # store messages_inflight separately to allow updating
+        ins.messages_inflight = messages_inflight
```
Comment (kbourgoin) [inline pykafka/producer.py]: If you have a mutable value, then an object is preferable to a namedtuple. Otherwise, it's not really a tuple anymore. If you're worried about the size of the object, you can use `__slots__`, but realistically there aren't going to be many of these hanging around, so it's probably not necessary.

(disclaimer: i realize protocol.py has this otherwise. It's something I've been working on and changed my mind on recently.)

File: pykafka/producer.py [hunk 12]
```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
...
         # a random partition distribution changing that on the retry.
         self._produce(self._partition_messages(messages), 0)
+
+
+_QueueInfo = namedtuple('QueueInfo', ['lock', 'flush_ready',
+                                      'slot_available', 'queue',
+                                      'producer'])
+
+
+class QueueInfo(_QueueInfo):
```
Comment (kbourgoin) [inline pykafka/producer.py]: It's not really "info" is it? It's really an async queue, since it contains the deque we're using as a queue.

File: pykafka/producer.py [hunk 13]
```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
...
+                messages_inflight):
+        ins = super(QueueInfo, cls).__new__(
+            cls, lock, flush_ready, slot_available, queue, producer)
+        # store messages_inflight separately to allow updating
+        ins.messages_inflight = messages_inflight
+        return ins
+
+
+class AsyncProducer():
+    """
```
Comment (kbourgoin) [inline pykafka/producer.py]: [nitpick] First line of the docstring should be a short description of the object, not a blank line.

File: pykafka/producer.py [hunk 14]
```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
...
+        # _workers_by_leader maps each broker id to a QueueInfo instance
+        # containing references to that broker's queue and its associated
+        # synchronization primitives
+        self._workers_by_leader = {}
+        for partition in self._topic.partitions.values():
+            if partition.leader.id not in self._workers_by_leader:
+                q_info = self._setup_queue(partition.leader)
+                self._workers_by_leader[partition.leader.id] = q_info
+
+    def _setup_queue(self, broker):
```
Comment (kbourgoin) [inline pykafka/producer.py]: The implementation no longer requires thread-per-broker because we already have a network thread for each broker. Adding more here just adds unnecessary complexity.

File: pykafka/producer.py [hunk 15]
```diff
@@ -241,3 +237,326 @@ def produce(self, messages):
...
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer"""
+        self.stop()
+
+    def _setup_internal_producer(self):
+        """Instantiate the internal producer"""
```
Comment (kbourgoin) [inline pykafka/producer.py]: Why an internal instance of the `Producer` instead of just subclassing?

File: pykafka/producer.py (hunk 14 above)
Comment (emmett9001) [inline pykafka/producer.py]: Talked about this with @kbourgoin, and it's not actually clear that using the existing network thread reduces any complexity.

File: pykafka/producer.py (hunk 9 above)
Comment (emmett9001) [inline pykafka/producer.py]: This poses an issue since we also need to use `Condition` variables here. I'm not sure how to integrate those into the handler.

File: pykafka/producer.py [hunk 16]
```diff
@@ -83,17 +75,34 @@ def __init__(self,
         :param retry_backoff_ms: The amount of time (in milliseconds) to
             back off during produce request retries.
         :type retry_backoff_ms: int
-        :param required_acks: How many other brokers must have committed the
-            data to their log and acknowledged this to the leader before a
-            request is considered complete?
+        :param required_acks: The number of other brokers must have committed
```
Comment (yungchin) [inline pykafka/producer.py]: Probably missing a "that" before "must"

File: pykafka/producer.py [hunk 17]
```diff
@@ -111,15 +125,114 @@ def __repr__(self):
...
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish - just re-raise
+        the exception. If __exit__ is being called because the with: block
+        has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        if self._worker_exception is not None:
+            log.exception(traceback)
+            self.stop(wait=False)
+            raise self._worker_exception
```
Comment (yungchin) [inline pykafka/producer.py]: I'm cheating and reading the context manager docs on the side, because I'd never written one, but those suggest this would be expected to `return False` instead of re-raising.

Comment (emmett9001) [inline pykafka/producer.py]: The reason I'm not doing that is that the exception that's raised when this condition is `True` is not the same exception held in `exc_value`.

Comment (yungchin) [inline pykafka/producer.py]: But wait - I'm probably confused. Is `self._worker_exception` always the same exception as the `exc_value` that caused the exit (that gets passed in args)?

Comment (yungchin) [inline pykafka/producer.py]: Ah, cross-post :)
Ok, got it!

Comment (emmett9001) [inline pykafka/producer.py]: `__exit__` is called when the `with` block ends or when an exception is raised. `self._worker_exception` being raised on a worker thread would not cause `__exit__` to be called. I just realized, though, that since all of the Producer's public methods raise `self._worker_exception` if it's set, `__exit__` doesn't actually need to worry about this exception at all.

File: pykafka/producer.py [hunk 18]
```diff
@@ -167,7 +292,8 @@ def _get_partition_msgs(partition_id, req):
                     to_retry.extend(_get_partition_msgs(partition, req))
```
Comment (yungchin) [inline pykafka/producer.py]: I know these lines didn't change in this pullreq, but I think it's worth flagging up that these produce-errors (the ones that bear a `continue`) only cause a log warning, so they're sort of silently dropped.

File: pykafka/producer.py [hunk 19]
```diff
@@ -111,15 +131,114 @@ def __repr__(self):
...
+        if wait:
+            self._wait_all()
+
+    def produce(self, messages):
+        """Produce a set of messages.
+
+        :param messages: The messages to produce
+        :type messages: Iterable of str or (str, str) tuples
+        """
+        def _partition_messages():
```
Comment (yungchin) [inline pykafka/producer.py]: Should this check `self._running`?

File: pykafka/producer.py [hunk 20]
```diff
@@ -237,7 +428,167 @@ def produce(self, messages):
         :param messages: The messages to produce
         :type messages: Iterable of str or (str, str) tuples
         """
-        # Do partition distribution here. We need to be able to retry producing
-        # only *some* messages when a leader changes. Therefore, we don't want
-        # a random partition distribution changing that on the retry.
-        self._produce(self._partition_messages(messages), 0)
+        super(SynchronousProducer, self).produce(messages)
+        self._wait_all()
```
Comment (yungchin) [inline pykafka/producer.py]: If this will be the only difference between the two producers, couldn't we just have one producer, with a `synchronous=True` kwarg on the `produce` method?

Comment (emmett9001) [inline pykafka/producer.py]: Yeah, maybe that's better. It would eliminate the need to duplicate all of the kwarg documentation, and I can't think of any downsides.

File: pykafka/producer.py [hunk 21]
```diff
@@ -237,7 +428,167 @@ def produce(self, messages):
...
+            if acquire:
+                self.lock.acquire()
+            if len(self.queue) >= self.producer._max_queued_messages:
+                self.slot_available.clear()
+            if acquire:
+                self.lock.release()
+            if self.producer._block_on_queue_full:
+                self.slot_available.wait()
+            else:
+                raise ProducerQueueFullError("Queue full for broker %d",
```
Comment (yungchin) [inline pykafka/producer.py]: When I call `produce` with a bunch of messages and this happens, is there an easy way to find out which of my messages got queued to an OwnedBroker and which didn't?

File: pykafka/producer.py [hunk 22]
```diff
@@ -237,7 +428,167 @@ def produce(self, messages):
...
+        :type acquire: bool
+        """
+        assert acquire or self.lock.locked()
+        self._wait_for_slot_available(acquire=acquire)
+        if acquire:
+            self.lock.acquire()
+        self.queue.extendleft(messages)
+        self.messages_inflight += len(messages)
+        if len(self.queue) >= self.producer._max_queued_messages:
+            if not self.flush_ready.is_set():
```
Comment (yungchin) [inline pykafka/producer.py]: In rdkafka there's one more parameter besides `_max_queued_messages` (which is called `min.queued.messages`, which is a bit of a confusing name perhaps), which I suspect might help performance: instead of waiting until you hit max_queued, at which point the `produce` call may block while waiting for the worker thread to flush, you allow the worker thread to flush as soon as min_queued is exceeded. I wouldn't know a good name for that parameter though :)

File: pykafka/producer.py (hunk 21 above)
Comment (emmett9001) [inline pykafka/producer.py]: Here's a potential solution. https://github.com/Parsely/pykafka/commit/597641040464cf5c7e327925b0c0ea6ab1eb509f To support infinite sequences, `produce` has to turn into a generator.

File: pykafka/producer.py (hunk 22 above)
Comment (emmett9001) [inline pykafka/producer.py]: Sounds like you're essentially suggesting that the message count thresholds for emptying the queue and declaring that it's full. Seems reasonable to me.

File: pykafka/producer.py [hunk 23]
```diff
@@ -17,17 +17,25 @@
 See the License for the specific language governing permissions and
 limitations under the License.
 """
-__all__ = ["Producer", "AsyncProducer"]
+__all__ = ["Producer", "SynchronousProducer"]
```
Comment (kbourgoin) [inline pykafka/producer.py]: Was going to comment about this being a breaking change, but it looks like this is how the JVM client is named too. However, I'd recommend `SyncProducer`. If Java has the short name, let's not out-length it.

File: pykafka/producer.py [hunk 24]
```diff
@@ -36,26 +44,16 @@
...
-        raise NotImplementedError("AsyncProducer is unimplemented")
-
-
-class Producer():
+class Producer(object):
     """
-    This class implements the synchronous producer logic found in the
-    JVM driver.
+    This class implements asynchronous producer logic similar to that found in
+    the JVM driver.
```
Comment (kbourgoin) [inline pykafka/producer.py]: [nitpick] Docstring first lines should be short, one-line, and describe the class. It make code folding and doc generation cleaner. It should also be on the same line as the triple quote.

```
"""Asynchronous Producer Implementation.

Blah blah blah.
"""
```

File: pykafka/producer.py [hunk 25]
```diff
@@ -102,7 +124,32 @@ def __init__(self,
...
+        self._linger_ms = linger_ms
+        self._block_on_queue_full = block_on_queue_full
+        self._worker_exception = None
+        self._worker_trace_logged = False
+        self._running = False
+        self.start()
+
+    def raise_worker_exceptions(fn):
+        """Decorator that raises exceptions encountered on worker threads
+        """
```
Comment (kbourgoin) [inline pykafka/producer.py]: [nitpick] If the docstring fits on one line, the closing triple-quote should be on that line too.

File: pykafka/producer.py [hunk 26]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
             id_=hex(id(self))
         )
 
-    def _send_request(self, broker, req, attempt):
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
```
Comment (kbourgoin) [inline pykafka/producer.py]: You're calling `self.start()` twice. In `__init__` and here. I'd leave it in `__init__` and remove this one.

File: pykafka/producer.py [hunk 27]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
...
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish. If __exit__ is being
+        called because the with: block has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        self.stop(wait=self._worker_exception is None)
```
Comment (kbourgoin) [inline pykafka/producer.py]: If there's an exception here, `wait` doesn't matter because it'll get raised before that matters.

File: pykafka/producer.py [hunk 28]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
...
+
+    @raise_worker_exceptions
```
Comment (kbourgoin) [inline pykafka/producer.py]: How is it possible there have been exceptions before `start` is called? It's called at the end of `__init__` before the workers have started.

File: pykafka/producer.py [hunk 29]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
...
+        """Set up data structures and start worker threads"""
+        if not self._running:
+            self._owned_brokers = {}
+            for partition in self._topic.partitions.values():
+                if partition.leader.id not in self._owned_brokers:
+                    self._owned_brokers[partition.leader.id] = OwnedBroker(
+                        self, partition.leader)
+            self._running = True
+
+    @raise_worker_exceptions
```
Comment (kbourgoin) [inline pykafka/producer.py]: I'm not sure this should raise exceptions, actually. If it raises because one worker is dead, how will the others be notified they should stop?

File: pykafka/producer.py [hunk 30]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
...
+                    key = None
+                    value = message
+                else:
+                    key, value = message
+                yield (key, str(value)), self._partitioner(partitions, message).id
+        if not self._running:
+            raise ProducerStoppedException()
+        pairs = itertools.izip(self._produce(_partition_messages()), messages)
+        for message_enqueued in pairs:
+            yield message_enqueued
```
Comment (kbourgoin) [inline pykafka/producer.py]: Why is produce giving you the messages right back? This seems unnecessary.

Comment (kbourgoin) [inline pykafka/producer.py]: And potentially confusing as to what a user should do with this.

File: pykafka/producer.py [hunk 31]
```diff
@@ -111,15 +158,112 @@ def __repr__(self):
...
+            request.add_message(
+                Message(value, partition_key=key),
+                self._topic.name,
+                partition_id
+            )
+        log.debug("Sending %d messages to broker %d",
+                  len(message_batch), owned_broker.broker.id)
+        self._send_request(request, 0, owned_broker)
+
+    @raise_worker_exceptions
```
Comment (kbourgoin) [inline pykafka/producer.py]: Doesn't this run on a background thread?

File: pykafka/producer.py [hunk 32]
```diff
@@ -102,7 +124,32 @@ def __init__(self,
...
+        self._max_queued_messages = max_queued_messages
+        self._min_queued_messages = min_queued_messages
+        self._linger_ms = linger_ms
+        self._block_on_queue_full = block_on_queue_full
+        self._worker_exception = None
+        self._worker_trace_logged = False
+        self._running = False
+        self.start()
+
+    def raise_worker_exceptions(fn):
```
Comment (kbourgoin) [inline pykafka/producer.py]: So, I'm actually not a fan of a decorator here.  This is a check which needs to happen, but it also looks like something which was sprinkled around without a ton of consideration as to when we're raising exceptions and how. It's called in places where it's detrimental (`stop`) and called places where it'll be ineffective (`_prepare_request` and `_send_request`). For the former case, it should be called after `self._running = False`, but you can't do that because it's a decorator.  In that latter case, it'll crash a background thread by raising an exception which won't be raised to the user and might kill an otherwise healthy thread.

This is why I don't like to decorator approach. It encourages adding it all over the place "just to be safe" because decorators look safe. Maybe it's just how I perceive it, but if this were a function call at the top of user-facing functions (i.e. a Guard), it'd be less likely to be added as it has been.

Maybe that's just me, though. Keep the decorator if you think it's useful, but right now it's used in way too many places. It should be used to guard user-facing functions, and in cases where it's not going to hurt things which otherwise should be happening (like shutting down the producer).

File: pykafka/producer.py [hunk 33]
```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
             attempt += 1
             if attempt < self._max_retries:
                 time.sleep(self._retry_backoff_ms / 1000)
-                self._produce(to_retry, attempt)
+                # we have to call _produce here since the broker/partition
+                # target for each message may have changed
+                for _ in self._produce(to_retry, attempt):
+                    pass
```
Comment (kbourgoin) [inline pykafka/producer.py]: This is definitely a point against having `_produce` return an iterator.

File: pykafka/producer.py (hunk 30 above)
Comment (kbourgoin) [inline pykafka/producer.py]: With everything being iterables here, what happens if someone calls `produce` and never iterates over the result?

File: pykafka/producer.py [hunk 34]
```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
...
-            else:
-                key, value = message
-            value = str(value)
-            yield (key, value), self._partitioner(partitions, message).id
+        # empty queues and figure out updated partition leaders
+        new_queue_contents = defaultdict(list)
+        for owned_broker in self._owned_brokers.itervalues():
+            owned_broker.lock.acquire()
+            current_queue_contents = owned_broker.flush(0, acquire=False,
+                                                        release_inflight=True)
```
Comment (kbourgoin) [inline pykafka/producer.py]: Use a re-entrant lock here. Having to tell the queue whether or not you want it to acquire the lock is poor separation of concerns. If you use the re-entrant lock, it'll be able to still acquire the lock since it's on the same thread.

File: pykafka/producer.py [hunk 35]
```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
...
-                key = None
-                value = message
-            else:
-                key, value = message
-            value = str(value)
-            yield (key, value), self._partitioner(partitions, message).id
+        # empty queues and figure out updated partition leaders
+        new_queue_contents = defaultdict(list)
+        for owned_broker in self._owned_brokers.itervalues():
+            owned_broker.lock.acquire()
```
Comment (kbourgoin) [inline pykafka/producer.py]: A better world is one where this function doesn't even know that the broker has locks. Is that possible here? I'm not actually sure.

File: pykafka/producer.py [hunk 36]
```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
...
+                                 self._block_on_queue_full,
+                                 acquire=False)
+            owned_broker.resolve_event_state()
+            owned_broker.lock.release()
 
-    def _produce(self, message_partition_tups, attempt):
-        """Publish a set of messages to relevant brokers.
+    @raise_worker_exceptions
+    def _wait_all(self):
+        """Block until all messages in flight are sent
```
Comment (kbourgoin) [inline pykafka/producer.py]: Just a thought, but "in flight" is an awkward term to use for these messages. When I see "in flight," I think "on the network on their way to the broker."  To me, this is more like "pending" messages since they're waiting to be sent. I can see how it cuts either way, so just a thought.

File: pykafka/producer.py (hunk 20 above)
Comment (kbourgoin) [inline pykafka/producer.py]: IIRC it's how the JVM producer is implemented. I have no strong feelings either way, and don't see any downsides to combining them.

File: pykafka/producer.py (hunk 32 above)
Comment (emmett9001) [inline pykafka/producer.py]: I've actually considered all of this, and have waffled back and forth between having it be a decorator and a simple two lines of code at the end of every function it's needed in. I liked the decorator approach better than calling a member function at the end of every function. I also considered only putting it on user-facing functions, but I chose to put it everywhere because of the possibility that users will call private functions. I couldn't find a case where having these checks on functions called on worker threads actually causes a problem - the thread will not have a chance to raise the exception a second time from the decorator, since it will already have exited after the original exception (from the `break` in its main loop error handler). I've also not been able to replicate a case where having `exit()` decorated causes a problem, but that's the one I'm least confident about.

Since a lot of what you've said here aligns with my gut instinct about this, I believe I'll keep the decorator on public methods only, unless I've inadvertently convinced you otherwise here.

File: pykafka/producer.py (hunk 36 above)
Comment (emmett9001) [inline pykafka/producer.py]: "pending" is much better, I don't think it's close.

File: pykafka/producer.py (hunk 34 above)
Comment (emmett9001) [inline pykafka/producer.py]: Thanks, this sounds like the "correct" way to do what I'm trying here.

File: pykafka/producer.py (hunk 32 above)
Comment (kbourgoin) [inline pykafka/producer.py]: The decorator's fine, but definitely only on the public functions. If someone is using private functions, that's their own problem.

As for background threads, this is using `self._worker_exception`, so as soon as that's set from any thread, doesn't that cause every other background thread to raise the exception once they call one a decorated function?

File: pykafka/producer.py (hunk 33 above)
Comment (emmett9001) [inline pykafka/producer.py]: Yeah, this is not great. Could use some wrapper function to make this code cleaner, and to automatically iterate the generator for users who want to ignore the yielded values.

File: pykafka/producer.py (hunk 31 above)
Comment (emmett9001) [inline pykafka/producer.py]: Yes, and I've come to believe that my reasoning for doing that was flawed. I'll just keep this decorator on public methods.

File: pykafka/producer.py (hunk 30 above)
Comment (emmett9001) [inline pykafka/producer.py]: After discussion, we agree that it makes more sense to stop supporting sequences of messages in `produce`. That will make the whole `block_on_queue_full` issue go away. It will also require that we use a major version bump.

File: pykafka/producer.py (hunk 29 above)
Comment (emmett9001) [inline pykafka/producer.py]: Agree, this will be removed.

File: pykafka/producer.py (hunk 28 above)
Comment (emmett9001) [inline pykafka/producer.py]: These exceptions would have happened during `start`, not before it, since `raise_worker_exceptions` does its work after running the decorated function.

File: pykafka/producer.py (hunk 23 above)
Comment (emmett9001) [inline pykafka/producer.py]: What do you think about completely removing the `SyncProducer` class and having synchronous behavior switched on and off by a kwarg on `Producer`?

Comment (kbourgoin) [inline pykafka/producer.py]: Yeah, I think that's a great idea. If we're in the real of making breaking changes, that'd make things overall a lot cleaner.

File: pykafka/producer.py (hunk 35 above)
Comment (yungchin) [inline pykafka/producer.py]: My guess would be that's not possible here, because you need to hold locks across all brokers, to prevent new messages getting slipped-in between the old ones while they are being redistributed.

File: pykafka/producer.py (hunk 27 above)
Comment (emmett9001) [inline pykafka/producer.py]: I don't follow this logic. Where is the exception guaranteed to be raised before the context manager exits? The idea here is to not hang infinitely waiting on the messages to send when the worker thread has crashed.

Comment (kbourgoin) [inline pykafka/producer.py]: You know, this is what happens when I don't read the decorator closely.  :)

I'm thought it did stuff before the function call, not after, so that wasn't necessarily true. Of course, the decorator is gone now, so the point is moot.

Comment (emmett9001) [inline pykafka/producer.py]: I got the impression you thought that, but wasn't certain. I'm not actually sure that there's a case where this setting will matter, but it's a safeguard in case there is one.

File: pykafka/producer.py (hunk 35 above)
Comment (emmett9001) [inline pykafka/producer.py]: Yeah I think the nature of the problem means that this function has to lock them all at once.

File: pykafka/producer.py [hunk 37]
```diff
@@ -111,15 +157,97 @@ def __repr__(self):
...
+        :param message: The message to produce
+        :type message: str
+        :param partition_key: The key to use when deciding which partition to send this
+            message to
+        :type partition_key: str
+        """
+        if not self._running:
+            raise ProducerStoppedException()
+        partitions = self._topic.partitions.values()
+        partition_id = self._partitioner(partitions, (message, partition_key)).id
```
Comment (yungchin) [inline pykafka/producer.py]: This bug reported in #232 affects this line too.

File: pykafka/producer.py [hunk 38]
```diff
@@ -111,15 +157,97 @@ def __repr__(self):
...
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish. If __exit__ is being
+        called because the with: block has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
```
Comment (yungchin) [inline pykafka/producer.py]: This comment and part of the docstring are leftovers from a previous iteration.

Comment (yungchin) [inline pykafka/producer.py]: Something else: do we also need a `__del__` that calls `stop` too?
